### ISSUE TITLE
na_ontap_quotas fails with API error 13001

### ISSUE THREAD

**Issue by mamoep** [ansible-collections/netapp.ontap#4]
##### SUMMARY

When creating or resizing tree quotas, the task sometimes fails with API error 13001. I don't fully understand the pattern, but it is reproducable for certain qtrees. I found a workaround setting the "policy" option. We only use the default quota policy, so we didn't use this option so far.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

na_ontap_quotas

##### ANSIBLE VERSION

```paste below
Ansible 2.9.19
Tower 3.8.2
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

ONTAP 9.7P12 & P13

##### STEPS TO REPRODUCE

Issue is not occuring for all qtrees. If a qtree is affected, the issue occurs everytime the playbook is run.

```yaml
  - name: Resize quotas
    na_ontap_quotas:
      state: present
      vserver: "{{ vserver_name }}"
      volume: "{{ volume_name }}"
      quota_target: "/vol/{{ volume_name }}/{{ qtree.name }}"
      type: tree
      disk_limit: "{{ qtree.size }}"
      set_quota_status: True
      activate_quota_on_change: resize
      <<: *login
    connection: local
```

##### EXPECTED RESULTS

Qtree quota should be changed to the desired size.

##### ACTUAL RESULTS

```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_na_ontap_quotas_payload_4r6TE2/ansible_na_ontap_quotas_payload.zip/ansible_collections/netapp/ontap/plugins/modules/na_ontap_quotas.py", line 287, in get_quotas
  File "/var/lib/awx/venv/ansible/lib/python2.7/site-packages/netapp_lib/api/zapi/zapi.py", line 301, in invoke_successfully
    raise NaApiError(code, msg)
NaApiError: NetApp API failed. Reason - 13001:success
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "activate_quota_on_change": "resize", 
            "cert_filepath": null, 
            "disk_limit": "10g", 
            "feature_flags": {}, 
            "file_limit": null, 
            "hostname": "cluster.domain.local", 
            "http_port": null, 
            "https": true, 
            "key_filepath": null, 
            "ontapi": null, 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "perform_user_mapping": null, 
            "policy": null, 
            "qtree": "", 
            "quota_target": "/vol/vol_name/foo-bar_auto1", 
            "set_quota_status": true, 
            "soft_disk_limit": null, 
            "soft_file_limit": null, 
            "state": "present", 
            "threshold": null, 
            "type": "tree", 
            "use_rest": "auto", 
            "username": "user", 
            "validate_certs": true, 
            "volume": "vol_name", 
            "vserver": "vserver"
        }
    }, 
    "msg": "Error fetching quotas info: NetApp API failed. Reason - 13001:success"
}
```

##### WORKAROUND
Using the same task and adding the policy option works for affected qtrees. Example:
```yaml
  - name: Resize quotas
    na_ontap_quotas:
      state: present
      vserver: "{{ vserver_name }}"
      volume: "{{ volume_name }}"
      policy: default
      quota_target: "/vol/{{ volume_name }}/{{ qtree.name }}"
      type: tree
      disk_limit: "{{ qtree.size }}"
      set_quota_status: True
      activate_quota_on_change: resize
      <<: *login
    connection: local
```

**Comment by lonico** [comment]
The error `Reason - 13001:success` is what we receive from ONTAP.  The success is a bit surprising.

This is how we're building the ZAPI request.  Since this is using `iter`, it should not be an issue if policy is missing.

```
        quota_get = netapp_utils.zapi.NaElement('quota-list-entries-iter')
        query = {
            'query': {
                'quota-entry': {
                    'volume': self.parameters['volume'],
                    'quota-target': self.parameters['quota_target'],
                    'quota-type': self.parameters['type'],
                    'vserver': self.parameters['vserver']
                }
            }
        }
        quota_get.translate_struct(query)
        if self.parameters.get('policy'):
            quota_get['query']['quota-entry'].add_new_child('policy', self.parameters['policy'])
```

**Comment by mamoep** [comment]
Is there a way to log the full ZAPI request that the module does? I think this would be helpful when I try to open an ONTAP support case.

**Comment by lonico** [comment]
In your task, you can add the following:
```
feature_flags:
    trace_apis: true
```
It will write the trace to /tmp/ontap_apis.log

**Comment by lonico** [comment]
Is there a chance that there is another policy on the qtree, other than default?  And that this policy has a rule that is malformed and breaks the ZAPI.    This sounds a bit far fetched.

**Comment by mamoep** [comment]
I am sure there is only the default policy.

**Comment by mamoep** [comment]
I compared the trace logs and the requests look like you described. I opened a support case.

**Comment by lonico** [comment]
Do you have a case number?

I think we could add a retry in the module if the first call fails, using 'default' is no policy was provided.   Though I don't want to hide some ONTAP issue, unless it's harmless.

**Comment by mamoep** [comment]
Case number is 2008802810. No findings so far. I don’t know if retrying with default policy is a valid fallback. I could think of cases were another policy name is the active one.

**Comment by lonico** [comment]
Thank you.  Yes, nothing interesting yet.   And I'm sure retrying is a good idea either.

**Comment by lonico** [comment]
I thought about doing something a bit more complicated.  If the ZAPI request fails with no policy specified:

1. get a list of policies
2. if there is only one policy, retry with this policy
3. if there are more, provide the policy details and fail.

**Comment by mamoep** [comment]
Support opened bug #1430725 for further investigation.

**Comment by jaiw23** [comment]
I kinda think found a pattern to this issue. If I try to add a new policy rule which is at the start of any existing quota rules (alphabetical order), you get this error. Example quota-policy = Default, quota-rules (lonserver1:25g  ,lonserver2:25g). If I try to add a new rule for (fraserver1:25g), it will give this error and fail. But If I add (nycserver1:25g) works fine. I suspect since fraserver1 would come before lonserver1 in (alphabetical) which is the start of the rule, it fails.

**Comment by lonico** [comment]
@jaiw23 Thank you for the update.  I was trying to use your example to replicate the issue, but it works for me.  Do you have a playbook you could share?  I may have skipped some important details.

**Comment by lonico** [comment]
fixed in 692fb756607b540e379bb2b8d8660bb04f5559a3

The fix provides a way to work around the underlying ONTAP issue.
